src/base/auto-complete/auto-complete-widget.ts

```diff
@@ -0,0 +1,304 @@
+/********************************************************************************
+ * Copyright (c) 2020 EclipseSource and others.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License v. 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ *
+ * This Source Code may also be made available under the following Secondary
+ * Licenses when the conditions for such availability set forth in the Eclipse
+ * Public License v. 2.0 are satisfied: GNU General Public License, version 2
+ * with the GNU Classpath Exception which is available at
+ * https://www.gnu.org/software/classpath/license.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ ********************************************************************************/
+import { AutocompleteResult, AutocompleteSettings } from "autocompleter";
+import { Action, ILogger, isAction, isLabeledAction, LabeledAction, SModelRoot } from "sprotty/lib";
+import { toArray } from "sprotty/lib/utils/iterable";
+import { matchesKeystroke } from "sprotty/lib/utils/keyboard";
+
+import { ValidationStatus } from "../actions/edit-validation-actions";
+import { SetAutoCompleteValueAction } from "./auto-complete-actions";
+import { IValidationDecorator } from "./validation-decorator";
+
+export interface AutoCompleteSettings {
+    readonly noSuggestionsMessage: string;
+    readonly suggestionsClass: string;
+    readonly debounceWaitMs: number;
+    readonly showOnFocus: boolean;
+}
+
+export abstract class InputValidator {
```

Comment:
Is there a particular reason why you use abstract classes instead of interfaces here?